### PR TITLE
[16.07] Fix metadata on HDAs in dynamically discovered collections.

### LINKED ISSUE #3267: Metadata not being set on dataset collection elements produced by tools
I'm running the tip of the dev branch, and I've recently noticed that metadata is not being generated for output dataset collection elements with tools that produce dataset collections on output.

Here is a simple tool that can be used to reproduce this issue.
```
<tool id="multi_output" name="Multi_Output" version="0.1.0">
  <command>
    <![CDATA[
        #import os
        mkdir -p output &&
        #for $i in range(2):
            #set ofile = 'output_%d.txt' % $i
            #set opath=os.path.join('output', $ofile)
            echo "Hello world" > $opath
        #end for
    ]]>
  </command>
  <inputs>
    <param name="input" type="integer" value="2" />
  </inputs>
  <outputs>
      <collection name="outs" type="list">
          <discover_datasets pattern="__name__" directory="output" visible="false" format="txt" />
      </collection>
  </outputs>
  <tests>
    <test>
    </test>
  </tests>
  <help>
    Help me
  </help>
  <citations />
</tool>
```
The history items that are elements of the collection will not display the expected info: or peek, and, with the exception of the 'dbkey: ?' entry, the metadata is missing in the database for the history_dataset_association records for each collection element.

### PR DESCRIPTION
**Pull request by jmchilton**
With test case.

Fixes #3267.

### CODE REVIEW

File: test/api/test_tools.py [hunk 1]
```diff
@@ -465,10 +465,14 @@ def test_dynamic_list_output( self ):
...
-        assert len( output_collection[ "elements" ] ) == 2
         self.assertEquals( output_collection[ "name" ], "Table split on first column" )
 
-        # TODO: verify element identifiers
+        assert len( output_collection[ "elements" ] ) == 2
+        output_element_0 = output_collection["elements"][0]
+        assert output_element_0["element_index"] == 0
+        assert output_element_0["element_identifier"] == "samp1"
+        output_element_hda_0 = output_element_0["object"]
+        assert output_element_hda_0["metadata_column_types"] != None
```
Comment (mvdbeek) [inline test/api/test_tools.py]: I think [flake8](https://travis-ci.org/galaxyproject/galaxy/jobs/189194240#L177) wants a `is not None` or `assert output_element_hda_0["metadata_column_types"]` here.

**Comment by mvdbeek** [comment]
Tests are passing locally, if you can fix the travis failure you got my :+1:

**Comment by jmchilton** [comment]
@galaxybot test this

**Comment by dannon** [comment]
Looks good to me.  Merging forward now.

**Comment by jmchilton** [comment]
Awesome sauce - thanks @dannon and @mvdbeek!

**Pull request merged**